Processing Parameters Extension #276 #545 (#471)
Open-EO/openeo-api: extensions/processing-parameters/README.md

# Processing Parameters Extension

The Processing Parameters Extension to the openEO API provides an interface to explore and handle additional processing options that a back-end can offer for the three processing modes (synchronous processing, batch jobs, secondary web services).

- Version: **0.1.0**
- Stability: **experimental**
- [OpenAPI document](openapi.yaml)
- Conformance class: `https://api.openeo.org/extensions/processing-parameters/0.1.0`

**Note:** This document only documents the additions to the specification.
Extensions can not change or break existing behavior of the openEO API.

This extension adds a new endpoint (`GET /processing_options`, see [OpenAPI document](openapi.yaml))
to discover the additional processing options that a back-end offers.

> Additionally, this extension allows to provide specific default values for user-defined processes (UDPs, see below).

Reviewer: I don't think this extension has to be limited to user-defined processes. It also should apply to remote process definitions (https://github.com/Open-EO/openeo-api/tree/draft/extensions/remote-process-definition).

Ultimately, it could also apply to predefined processes. However, as mentioned at https://github.com/Open-EO/openeo-api/issues/545#issuecomment-2362487544), this would  be kind of an anti-pattern against the vision of openEO, so I don't think this should be (explicitly) in scope at this point.

```suggestion
Additionally, this extension allows to provide specific default values for processing values, embedded in process definitions, such as user-defined processes (UDPs, see below) and [remote process definitions](https://github.com/Open-EO/openeo-api/tree/processing-parameters/extensions/processing-parameters).
```

Author: Remote process definitions are UDPs, too. I don't think I understand why the differentiation is needed.
Process definition is not a term in the API spec, but UDP is. So I'm inclined to leave it as is.

Author: Nevertheles, I clarified that there are various "sources" for such UDPs, including remote UDPs:
https://github.com/Open-EO/openeo-api/pull/471/commits/f7b765bf71272420e62019340102c08b111df5d9

Reviewer: ok fair enough. For me, the term "UDP" in the openEO API spec was always closely tied to the `/process_graphs` endpoints. But indeed, after checking in more detail, the openEO API  is a bit more relaxed about the term.

In both cases, the parameters and its values are provided separately for each processing mode.

> ## Processing Parameters in UDPs

Reviewer: ```suggestion
## Embedding default processing options in process definitions
```

Author: ```suggestion
## Embedding default processing options in UDPs
```

> UDPs can provide default values for specific processing parameters.

Reviewer: ```suggestion
Process definitions (such as user-defined processes or remote process definition resources) can include default values for specific processing parameters.
```

The values for each parameter are provided  separately for each processing mode.

> The following properties are added to the top-level of a UDP for the respective processing modes:

Reviewer: ```suggestion
The following properties are added to the top-level of a process definition (e.g. as sibling nodes to the "process_graph" property) for the respective processing modes:
```

Author: ```suggestion
The following properties are added to the top-level of a UDP (e.g. as sibling nodes to the "process_graph" property) for the respective processing modes:
```

- `default_synchronous_parameters` for synchronous processing
- `default_job_parameters` for batch jobs

> - `default_service_parameters` for secondary web services

Reviewer: (feel free to ignore if this is too much bike-shedding)

I'm not sure if it is intentional and there is a reasoning behind it, but the current description mixes "parameter" and "option": "processing parameter extension", `GET /processing_options`, `default_job_parameters`, ...

To me, "parameter" sounds more like a non-concrete placeholder, while "option" sounds a bit more concrete (e.g. a value provided by the user to make a parameter concrete). So in that sense, I'd think that the naming should be switched to:
- `GET /processing_parameters`
- `default_synchronous_options`,  `default_job_options`, ....

I would also be fine with just collapsing everything to a single name, e.g. "option"

Author: Indeed, it should just be a single name for it. Not sure yet which is better, but something to consider. Will have another look tomorrow.

Reviewer: Can we proceed on picking a single name for this concept? We're starting on implementing support for this, so it would be best if we can settle this detail.

I tend to prefer "option".
"Parameter" is already a defined concept in openeo process graph handling (including resolution and scoping rules), so I would avoid overloading it here with a somewhat different/incompatible meaning.

Author: I was thinking into the same direction, but we have also named it paramerter pretty much everywhere else in the API, in services types, in file formats, ... (but we use options in the processes). In the programming clients they'd probably be exposed as parameters? Or as a single parameter called options?

Not sure, any other preferences?

Reviewer: > file formats, ... (but we use options in the processes). In the programming clients they'd probably be exposed as parameters? Or as a single parameter called options?

Interesting I didn't notice that yet. In the openEO API, under `GET /file_formats` we speak about "parameters" (including the "parameter" property in a file format object), but in `save_result` we use argument name `options` (while its description still refers to "parameters" however). Was that an intended choice or did it just grow organically to that?

(FYI: in the Python client, we mainly expose the `options` argument of `save_result`, there is not really a prominent reference to "parameters")

Reviewer: > named it paramerter pretty much everywhere else in the API, in services types

under service types there is indeed "process_parameters", but that is directly tied to the classic parameters we have in process graph handling, so that's fine I think

Reviewer: The existing situation of file formats is an interesting precedent actually. While sticking to a single name might be easier for end users to connect the dots, it also makes sense to be bit more nuanced about the difference between defining a parameter and providing an actual value for it, e.g.
- the backend provides a file format with a _parameter_  (a non-concrete placeholder) for example "jpeg_quality"
- the user picks  _option_  `80%` (a concrete value) for jpeg quality

If we want to align with that existing pattern, then it would be here I think:
- name "Processing parameter extension"  (already the case in current PR)
- new endpoint `GET /processing_parameters`   (this is different)
- default values in UDPs and related: `default_synchronous_options`, `default_job_options`, `default_service_options`  (also different)

Author: That makes sense to me.

Author: See https://github.com/Open-EO/openeo-api/pull/471/commits/e9ed2bbcd0b32a5007523cef1433d272ef1dea9a

The schema for each of these properties is:

```yaml
type: object
additionalProperties:
  description: Any type
```

> The keys of the object are the respecive parameter names.

Reviewer: ```suggestion
The keys of the object are the respective parameter names.
```

> The values of the object are the (default) values for the parameters.

Reviewer: ```suggestion
The values of the object are the default values for the parameters.
```
I don't think there is a reason to put "default" between parentheses

Schematic restrictions are not defined for the object, but the schemas for the parameters as defined in `GET /processing_options` apply to the given values.
These values provide the defaults unless a user overrides them in the actual data processing request (e.g. `POST /jobs`, see below).

> 

Reviewer: Maybe add an example here to make things a bit more concrete:

```
Example process definition including processing option defaults:

{
    "id": "my_evi",
    "parameters": [...],
    "process_graph": {...},
    "default_job_parameters": {
        "memory": "2GB",
        "block-sizes": [128, 32]
    }
}
```

Author: See https://github.com/Open-EO/openeo-api/pull/471/commits/a407db56dcb22633f218622d547f7c97dc61f057

## Resolving parameters

Due to the variety of places where processing parameters can be provided, the following
list defines how the parameters must be resolved. The prioritization is as follows:

1. If present, use the parameter specified in the processing request directly (e.g. in `POST /jobs` as a top-level property)

> 2. If present, use the default parameter specified in the UDP

Reviewer: ```suggestion
2. If present, use the default parameter specified in the process definition
```

# Processing Parameters Extension

The Processing Parameters Extension to the openEO API provides an interface to explore and handle additional processing parameters that a back-end can offer for the three processing modes (synchronous processing, batch jobs, secondary web services).

- Version: **0.1.0**
- Stability: **experimental**
- [OpenAPI document](openapi.yaml)
- Conformance class: `https://api.openeo.org/extensions/processing-parameters/0.1.0`

**Note:** This document only documents the additions to the specification.
Extensions can not change or break existing behavior of the openEO API.

This extension adds a new endpoint (`GET /processing_parameters`, see [OpenAPI document](openapi.yaml))
to discover the additional processing parameters that a back-end offers.

Additionally, this extension allows to provide specific default values for user-defined processes (UDPs),
which includes:

> - UDPs submitted directly for synchronouzs processing, as batch jobs, or as secondary webservices

Reviewer: ```suggestion
- UDPs submitted directly for synchronous processing, as batch jobs, or as secondary web services
```
